src/main/scala/chisel3/experimental/inlinetest/InlineTest.scala

```diff
@@ -114,13 +139,28 @@ trait HasTests[M <: RawModule] { module: M =>
     */
   protected final def test[R](
     testName: String
-  )(testBody: Instance[M] => R)(implicit th: TestHarnessGenerator[M, R]): Unit =
-    elaborateParentModule { moduleDefinition =>
-      val resetType = module match {
-        case module: Module => Some(module.resetType)
-        case _ => None
-      }
-      val test = new TestParameters[M, R](desiredName, testName, moduleDefinition, testBody, resetType)
-      th.generate(test)
+  )(testBody: Instance[M] => R)(implicit testHarnessGenerator: TestHarnessGenerator[M, R]): Unit = {
+    val resetType = module match {
+      case module: Module => Some(module.resetType)
+      case _ => None
     }
+    require(!testGenerators.contains(testName), s"test '${testName}' already declared")
+    val testGenerator =
+      new TestGenerator(
+        index = testGenerators.size,
+        testName,
+        () => module.name,
+        testBody,
+        resetType,
+        testHarnessGenerator
+      )
+    testGenerators.addOne(testName -> testGenerator)
```

Comment:
Nit: using the `+=` alias for `addOne` is a bit more common in Chisel internals:

```suggestion
    testGenerators += testName -> testGenerator
```